### DISCUSSION TITLE
How can I measure source voltage using ATTINY804.

### DISCUSSION THREAD

**Discussion by krishank652** [SpenceKonde/megaTinyCore#1081] [Q&A] [answered]
Hello,

I'm currently encountering a challenge in developing a program to monitor the supplied voltage in my circuit. The circuit features an ATTINY804 microcontroller and is powered by three AA type cells, resulting in an initial total voltage of 4.5V (1.5V per cell). Over time, the power level of the circuit gradually decreases, and I'm seeking a method to measure the voltage of the VCC.

My objective is to implement an alarm system that triggers when the voltage drops below 3V. Could you kindly assist me in providing a code snippet that enables the ATTINY804 MCU to read the supplied voltage accurately?

**Comment by hmeijdam** [comment, marked as answer]
This will print the actual VCC in the serial monitor.

```
void setup() {
Serial.begin(9600);
}

void loop() {
Serial.print("VCC: ");
Serial.println (readVDD());
delay(1000);
}

uint16_t readVDD(void){ //returns Vdd volts*100 (3.21v = return 321)
    //adc0 VREF = 1v1
    VREF.CTRLA = (VREF.CTRLA & ~VREF_ADC0REFSEL_gm) | VREF_ADC0REFSEL_1V1_gc;
    ADC0.CTRLB = ADC_SAMPNUM_ACC64_gc; //64 samples
    ADC0.CTRLC = ADC_REFSEL_VDDREF_gc | ADC_PRESC_DIV16_gc | ADC_SAMPCAP_bm;
    ADC0.CTRLD = ADC_INITDLY_DLY64_gc;
    ADC0.MUXPOS = ADC_MUXPOS_INTREF_gc;
    ADC0.CTRLA = 1; //enable adc
    ADC0.COMMAND = 1; //start
    while( ADC0.COMMAND ){} //wait
    return (1023*1.1*100ul*64) / ADC0.RES;
}
```

**Reply by krishank652** [reply]
@hmeijdam thanks, 
Its working for me.

 I also tried example : [megaavr/libraries/megaTinyCore/examples/readTempVcc/readTempVcc.ino](https://github.com/SpenceKonde/megaTinyCore/blob/master/megaavr/libraries/megaTinyCore/examples/readTempVcc/readTempVcc.ino)

both are working for me. Thanks I found the solution.
